lib/ash/subject.ex

```diff
@@ -0,0 +1,539 @@
+defmodule Ash.Subject do
+  @moduledoc """
+  Provides a consistent API for common operations across `Ash.Changeset`,
+  `Ash.Query`, and `Ash.ActionInput`. It allows you to write generic code that works
+  with any of these types without needing to pattern match or special-case your logic.
+  """
+
+  @type t :: Ash.Changeset.t() | Ash.Query.t() | Ash.ActionInput.t()
+
+  @doc """
+  Adds an error or list of errors to the subject.
+
+  Supports all subject types (Changeset, Query, ActionInput) and maintains
+  type consistency.
+
+  ## Parameters
+
+    * `subject` - The subject to add errors to
+    * `errors` - Error or list of errors to add
+    * `path` - Path where the error occurred (optional, defaults to [])
+  """
+  @spec add_error(t(), Ash.Error.error_subject() | list(Ash.Error.error_subject())) :: t()
+  @spec add_error(
+          t(),
+          Ash.Error.error_subject() | list(Ash.Error.error_subject()),
+          Ash.Error.path_subject()
+        ) :: t()
+  def add_error(subject, errors, path \\ [])
+
+  def add_error(subject, [], _path), do: subject
+
+  def add_error(%Ash.Changeset{} = subject, errors, path) when is_list(errors) do
+    Ash.Changeset.add_error(subject, errors, path)
+  end
+
+  def add_error(%Ash.Query{} = subject, errors, path) when is_list(errors) do
+    Ash.Query.add_error(subject, path, errors)
+  end
+
+  def add_error(%Ash.ActionInput{} = subject, errors, path) when is_list(errors) do
+    Ash.ActionInput.add_error(subject, errors, path)
+  end
+
+  def add_error(%Ash.Changeset{} = subject, error, path) do
+    Ash.Changeset.add_error(subject, error, path)
+  end
+
+  def add_error(%Ash.Query{} = subject, error, path) do
+    Ash.Query.add_error(subject, path, error)
+  end
+
+  def add_error(%Ash.ActionInput{} = subject, error, path) do
+    Ash.ActionInput.add_error(subject, error, path)
+  end
+
+  @doc """
+  Sets the actor for the subject.
+
+  The actor represents the user or entity performing the action.
+  Can be set directly or via keyword list with `:actor` key.
+
+  ## Parameters
+
+    * `subject` - The subject to set actor on
+    * `actor_or_opts` - Either the actor directly, or keyword list with `:actor` key
+  """
+  @spec set_actor(t(), Keyword.t() | term()) :: t()
+  def set_actor(subject, opts) when is_list(opts) do
+    if Keyword.has_key?(opts, :actor) do
+      set_actor(subject, opts[:actor])
+    else
+      subject
+    end
+  end
+
+  def set_actor(subject, actor) do
```

Comment:
AFAIK this is a private function on `Ash.Query` for example. `set_actor|tenant|authorize?|tracer` should be removed here.

Reply:
not private w/ `defp` but with `@doc false`